File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_87_evaluate_division.ts
function calcEquation(equations: string[][], values: number[], queries: string[][]): number[] {
    // Build graph: adjacency list with weights
    const graph = new Map<string, Map<string, number>>();
    
    for (let i = 0; i < equations.length; i++) {
        const [dividend, divisor] = equations[i];
        const value = values[i];
        
        if (!graph.has(dividend)) {
            graph.set(dividend, new Map<string, number>());
        }
        if (!graph.has(divisor)) {
            graph.set(divisor, new Map<string, number>());
        }
        
        graph.get(dividend)!.set(divisor, value);
        graph.get(divisor)!.set(dividend, 1.0 / value);
    }
    
    function bfs(start: string, end: string): number {
        // Check if variables exist in graph
        if (!graph.has(start) || !graph.has(end)) {
            return -1.0;
        }
        
        // If start equals end and it exists in graph
        if (start === end) {
            return 1.0;
        }
        
        // BFS to find path from start to end
        const queue: [string, number][] = [[start, 1.0]]; // [node, accumulated_product]
        const visited = new Set<string>([start]);
        
        while (queue.length > 0) {
            const [node, product] = queue.shift()!;
            
            // Check all neighbors
            const neighbors = graph.get(node)!;
            for (const [neighbor, weight] of neighbors.entries()) {
                if (neighbor === end) {
                    return product * weight;
                }
                
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    queue.push([neighbor, product * weight]);
                }
            }
        }
        
        return -1.0;
    }
    
    // Process all queries
    const results: number[] = [];
    for (const [dividend, divisor] of queries) {
        results.push(bfs(dividend, divisor));
    }
    
    return results;
}

File: src/my_project/interviews/top_150_questions_round_22/ex_87_evaluate_division.py
from typing import List, Union, Collection, Mapping, Optional
from abc import ABC, abstractmethod
from collections import deque, defaultdict

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        # Build graph: adjacency list with weights
        graph = defaultdict(dict)
        
        for (dividend, divisor), value in zip(equations, values):
            graph[dividend][divisor] = value
            graph[divisor][dividend] = 1.0 / value
        
        def bfs(start: str, end: str) -> float:
            # Check if variables exist in graph
            if start not in graph or end not in graph:
                return -1.0
            
            # If start equals end and it exists in graph
            if start == end:
                return 1.0
            
            # BFS to find path from start to end
            queue = deque([(start, 1.0)])  # (node, accumulated_product)
            visited = {start}
            
            while queue:
                node, product = queue.popleft()
                
                # Check all neighbors
                for neighbor, weight in graph[node].items():
                    if neighbor == end:
                        return product * weight
                    
                    if neighbor not in visited:
                        visited.add(neighbor)
                        queue.append((neighbor, product * weight))
            
            return -1.0
        
        # Process all queries
        results = []
        for dividend, divisor in queries:
            results.append(bfs(dividend, divisor))
        
        return results